Lambdas/GHOAuth/OAuthLambda.swift

```diff
@@ -0,0 +1,124 @@
+import DiscordBM
+import Foundation
+import SotoSecretsManager
+import GHHooksLambda
+import Models
+
+@main
+struct GHOAuthHandler: LambdaHandler {
+    typealias Event = APIGatewayV2Request
+    typealias Output = APIGatewayV2Response
+
+    let client: HTTPClient
+    let logger: Logger
+    let secretsRetriever: SecretsRetriever
+    let discordClient: any DiscordClient
+
+    init(context: LambdaInitializationContext) async throws {
+        self.client = HTTPClient(eventLoopGroupProvider: .shared(context.eventLoop))
+        self.logger = context.logger
+
+        let awsClient = AWSClient(httpClientProvider: .shared(client))
+        self.secretsRetriever = SecretsRetriever(awsClient: awsClient, logger: logger)
+
+        let botToken = try await secretsRetriever.getSecret(arnEnvVarKey: "BOT_TOKEN_ARN")
+        self.discordClient = await DefaultDiscordClient(httpClient: client, token: botToken)
+    }
+
+    func handle(_ event: APIGatewayV2Request, context: LambdaContext) async throws -> APIGatewayV2Response {
+        do {
+            return try await handleThrowing(event, context: context)
+        } catch {
+            do {
+                try await discordClient.createMessage(
+                    channelId: OAuthLambdaConstants.Channels.logs.id,
+                    payload: .init(embeds: [.init(
+                        title: "GHOAuth lambda top-level error",
+                        description: String("\(error)".prefix(4_000)),
+                        color: .red
+                    )])
+                ).guardSuccess()
+            } catch {
+                logger.error("DiscordClient logging error", metadata: [
+                    "error": "\(error)"
+                ])
+            }
+            throw error
+        }
+    }
+
+    func handleThrowing(
+        _ event: APIGatewayV2Request,
+        context: LambdaContext
+    ) async throws -> APIGatewayV2Response {
+        logger.trace("Received event: \(event)")
+
+        guard let code = event.queryStringParameters?["code"] else {
+            return .init(
+                statusCode: .badRequest,
+                body: "Missing code query parameter"
+            )
+        }
+
+        logger.trace("Retrieving secrets")
+        let clientSecret = try await secretsRetriever.getSecret(arnEnvVarKey: "GH_CLIENT_SECRET_ARN")
+        guard let clientID = ProcessInfo.processInfo.environment["GH_CLIENT_ID"] else {
+            throw OAuthLambdaError.envVarNotFound(name: "GH_CLIENT_ID")
+        }
+
+        // https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps
+
+        logger.trace("Requesting GitHub access token")
+        var request = HTTPClientRequest(url: "https://github.com/login/oauth/access_token")
+        request.method = .POST
+        request.headers = [
+            "Accept": "application/json",
+            "Content-Type": "application/json"
+        ]
+        let requestBody = try JSONEncoder().encode([
```

Comment:
@gwynne said `JSONEncoder()` and `JSONDecoder()` might have initialization costs, so better to move all decoder/encoders to be a property of this struct so they are only initialized once, like:
```swift
let jsonDecoder = JSONDecoder()
let jsonEncoder = JSONEncoder()
```
then use those in functions where needed.